Add phone number addressing proposal (#145)
jamesmunns/ergot: notes/proposals/0145-phone-number-addressing.md

* Their "Pool" of addresses that they can allocate FROM
* Their list of "Allocations", or address ranges within their Poll that have been exclusively assigned. These Allocations include:
    * Specific allocations, e.g. single-address assignments to sockets
    * Range allocations, e.g. for downstream devices

For example:

* A single device might allocate a range of `^7` to itself, to use as a range of socket addresses.
* A routing device may assign `^10` ranges to each of it's "downstream" connected devices

It is not currently specified whether the "Pool" of addresses are always contiguous, or a single `^N` allocation.

### Apex Entities

An entity without an "upstream" or "parent" entity above them is referred to as an "Apex" entity.

An Apex entity is able to use the entire 32-bit address space as it sees fit, as it has no upstream network that it needs to "fit within". However, Apex entities may still want to be conservative in how they use this space for two reasons:

1. The existence as an "Apex" may be temporary, if they later join a network as a "downstream" or "child" entity

> 2. The smaller addresses uses, the more compact they will be when transmitting.

Reviewer: ```suggestion
2. The smaller the addresses they use, the more compact those will be when transmitting.
```

> Currently, it is suggested that Apex entity make a best guess at the maximum address space needed for all downstream entities, including its own local sockets.

Reviewer: ```suggestion
Currently, it is suggested that Apex entities make a best guess at the maximum address space needed for all downstream entities, including their own local sockets.
```

### Downstream Entities

> Devices that are connected to another device are referred to as "Downstream" entities. They may be directly connected to an Apex entity, or to another Downstream entity with or more "hops" to the Apex entity.

Reviewer: ```suggestion
Devices that are connected to another device are referred to as "Downstream" entities. They may be directly connected to an Apex entity, or to another Downstream entity with one or more "hops" to the Apex entity.
```

3. C delivers to it's own socket range, as well as to D, both with the source address `183^9`. Let's say neither respond.
4. C's profile knows it's base address is `180^10`. It forwards to its parent, A, and updates the source address to `303^10`.
5. A delivers to it's own socket range, as well as entities B and F. Let's say only B cares, and it is delivered to B's 09^8.
6. B receives the discovery request, and notes that it has an `$LED_OUTPUT` service on socket `0A^8`, which it does ??? to determine that `0A^8` maps to `08A^10`. It replies TO `303^10` with this info.
7. B sends it to A:303^10 is outside it's range of 080^10 to 17F^10.
8. A sends it to C, as 303^10 is in the range 180^10 to 37F^10.
9. C sends it to E, as 303^10 is in the range 300^10 to 33F^10
10. E gives it to it's own sockets, as 303^10 is in the range 300^10 to 33F^10

Now: E can send messages to B, using `303^10` as the source (it got this in the header of the reply), and `08A^10` as the destination (it got this in the body of the reply). B can reply to E using the same.

As descrived, this generally requires:

1. All devices know their own range (e.g. E's socket knows it is `03^6`)

> 2. The profile of a device knows it's parents range, and it's own offset in that range (e.g. E's profile knows it is 180^9, so its socket is `183^9`

Reviewer: ```suggestion
2. The profile of a device knows its parent's range, and its own offset in that range (e.g. E's profile knows it is 180^9, so its socket is `183^9`
```
